src/ConfigFileParser.cpp

```diff
@@ -0,0 +1,534 @@
+
+    return m_configFile;
+}
+
+/**
+ * @brief Initializes the ConfigServer object with default values
+ *
+ * @param configServer The ConfigServer object to initialize
+ */
+void ConfigFileParser::initializeConfigServer(ConfigServer& configServer)
+{
+    const char* validServerDirectiveNames[] = { "server_name", "listen", "host", "client_max_body_size", "error_page", "location", "root" };
+    const int validServerDirectiveNamesSize = sizeof(validServerDirectiveNames) / sizeof(validServerDirectiveNames[0]);
+
+    configServer.validServerDirectives = std::vector<std::string>(validServerDirectiveNames, validServerDirectiveNames + validServerDirectiveNamesSize);
+    configServer.serverName = "";
+    configServer.root = "html";
+    configServer.listen.insert(std::make_pair("127.0.0.1", 80));
+    configServer.locationIndex = 0;
+    configServer.maxBodySize = 1;
+    configServer.errorPage = std::map<unsigned short, std::string>();
+    configServer.locationIndex = 0;
+    configServer.locations = std::vector<Location>();
+
+    m_configFile.servers.push_back(configServer);
+}
+
+/**
+ * @brief Initializes the Location object with default values
+ *
+ * @param location The Location object to initialize
+ */
+void ConfigFileParser::initializeLocation(Location& location)
+{
+    const char* validLocationDirectiveNames[] = { "root", "index", "cgi_ext", "cgi_path", "autoindex", "limit_except", "location", "return" };
+    const int validLocationDirectiveNamesSize = sizeof(validLocationDirectiveNames) / sizeof(validLocationDirectiveNames[0]);
+
+    location.path = "";
+    location.root = "html";
+    location.indices = std::vector<std::string>();
+    location.cgiExt = "";
+    location.cgiPath = "";
+    location.isAutoindex = false;
+    location.returns = std::map<unsigned short, std::string>();
+    location.validLocationDirectives = std::vector<std::string>(validLocationDirectiveNames, validLocationDirectiveNames + validLocationDirectiveNamesSize);
+    m_configFile.servers[m_configFile.serverIndex].locations.push_back(location);
+}
+
+/**
+ * @brief Checks if there are no open brackets
+ *
+ * If a opening bracket is found, it is pushed onto the brackets stack
+ * If a closing bracket is found, it is popped from the brackets stack
+ *
+ * At the end the stack should be empty. If not, there are open brackets
+ *
+ * @param configFilePath Path from the config file
+ * @return true If there is minimum one open bracket
+ * @return false If there no open bracket
+ */
+bool ConfigFileParser::isBracketOpen(const std::string& configFilePath) const
+{
+    std::ifstream tmpStream;
+    std::string tmpLine;
+    std::stack<char> brackets;
+
+    tmpStream.open(configFilePath.c_str());
+
+    while (getline(tmpStream, tmpLine)) {
+        for (std::string::const_iterator it = tmpLine.begin(); it != tmpLine.end(); ++it) {
+            if (*it == '{')
+                brackets.push('{');
+            else if (*it == '}' && brackets.empty())
```

Comment:
Yes, I misread.

I tried it out with TEST_F(ValidConfigFileTests, ValidFile), testing valid_config.conf. I adapted the config file with }}} and the test passed...

If I try to parse it in main, it throws as expected.

Reply:
So resolved? If yes, I suggest you do it, as you also opened the conversation - would follow this as a general rule: 

person who opens a conversation also resolves it.

Reply:
The whole ValidConfigFileTests suite is strange: no matter how you alter the config files, the tests always pass. 

Even when the valid_config.conf is empty, the test passes.

Reply:
Yeah I left it open because the tests behave strange. I think maybe because of file opening?

Reply:
I remember you guys had some issues with the tests. What was it about?

Reply:
It reported error, even though it shouldn't fail: the expectation was a throw, the config file should throw, but it didn't receive a throw.

We thought it had something to do with the test macro, because after a change it passed (as expected).

I know suspect that maybe the file opening and reading is a bit fragile? But even then it should throw (because file not opened) and therefore fail the test.

Maybe some other stuff is interfering (e.g. compiler optimisations?).

Needs investigation

Reply:
Thank you both very much for pointing this out!

I rewrote the checks for valid files in this manner:

```cpp
TEST_F(ValidConfigFileTests, ValidFile)

{
	EXPECT_NO_THROW(m_configFileParser.parseConfigFile("config_files/valid_config.conf"));
}
```

When now an exception gets thrown, the tests fail. 

Before I did this I used a lambda function for calling the parseConfigFile function. This was the case because I used this function before for testing invalid files and to check the exception message. 

This didn't work as expected, therefore I rewrote the checks for invalid files. But I still used the lambda functions in the functions for checking valid files. This was a mistake.